pkg/provider/notebook_resource.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	var ok bool
+	if r.provider, ok = req.ProviderData.(*Provider); !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *Provider, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+	}
+}
+
+func (r *NotebookResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan NotebookResourceModel
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get use case ID if specified
+	var useCaseID string
+	if !plan.UseCaseID.IsNull() {
+		useCaseID = plan.UseCaseID.ValueString()
+	}
+
+	// Read file
+	filePath := plan.FilePath.ValueString()
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		resp.Diagnostics.AddError("Error reading file", fmt.Sprintf("Unable to read file at %s: %s", filePath, err))
+		return
+	}
+
+	// Calculate file hash
+	hash := sha256.Sum256(content)
+	hashStr := hex.EncodeToString(hash[:])
+
+	fileName := filepath.Base(filePath)
+
+	traceAPICall("ImportNotebookFromFile")
+	importResp, err := r.provider.service.ImportNotebookFromFile(ctx, fileName, content, useCaseID)
+	if err != nil {
+		resp.Diagnostics.AddError("Error importing notebook", err.Error())
+		return
+	}
+
+	// Update model
+	plan.ID = types.StringValue(importResp.ID)
+	plan.Name = types.StringValue(importResp.Name)
+	plan.FileHash = types.StringValue(hashStr)
+
+	// Save model into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *NotebookResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state NotebookResourceModel
+
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	traceAPICall("GetNotebook")
+	notebook, err := r.provider.service.GetNotebook(ctx, state.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error reading notebook", err.Error())
+		return
+	}
+
+	// Update resource state with the current state
+	state.ID = types.StringValue(notebook.ID)
+	state.Name = types.StringValue(notebook.Name)
+
+	// Store the UseCaseID if available.
+	if notebook.UseCaseID != "" {
+		state.UseCaseID = types.StringValue(notebook.UseCaseID)
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *NotebookResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan, state NotebookResourceModel
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Handle use case updates.
+	if !state.UseCaseID.IsNull() && plan.UseCaseID.IsNull() {
+		// Remove notebook from old use case if it exists and re-import as a classic notebook.
```

Comment:
So apologies if I'm generally misunderstanding goals/expectations here.

I think this comment should change to something like:
```suggestion
		// Delete notebook and re-import as a classic notebook.
```

But, that being said, what is the aim here?
A user wants to "update" a notebook but pass no `UseCaseID` and then expect to "become" a classic notebook?
It really will be a whole new entity with a different ID (from the "original") as well.

Maybe instead what the flow should be is:
1. Download notebook (in its current state) as an .ipynb file
2. Create classic notebook with this file
3. Delete original notebooks (that is part of next-gen/in a Use Case)

Happy to discuss more - I feel I could be confusing things here.

Reply:
Now that I've added support for the local notebook changing on disk, I follow this flow both to convert things to classic notebooks and to handle re-importing the Notebook. It feels even more odd, so maybe we should catch up on going back to the drawing board here.

>A user wants to "update" a notebook but pass no UseCaseID and then expect to "become" a classic notebook?
It really will be a whole new entity with a different ID (from the "original") as well.

I mean these are TF semantics. I don't have to care as the user whether it is classic or not, I'm just modifying the state file, and terraform spits back a URL to get to it. If we don't want to allow that to happen, that's a different thing and I can make use case mandatory so that we don't have to deal with this edge case. It still won't help me with replacing the existing notebook though, so I'm not sure how much we gain by dropping support for this being optional.